torchgen/api/autograd.py

```diff
@@ -305,6 +306,164 @@ def dispatch_strategy(fn: NativeFunctionWithDifferentiabilityInfo) -> str:
         return "use_type"
 
 
+def is_foreach_func(f: NativeFunction) -> bool:
+    base_op_name = f.func.name.name
+    return base_op_name.base.startswith("_foreach_") and not base_op_name.inplace
+
+
+def is_reference_for_foreach(
+    f: NativeFunction,
+    function_schema: FunctionSchema,
+) -> bool:
+    return (
+        f.func.name.name.base.split("_foreach_")[-1] == function_schema.name.name.base
+        and not function_schema.name.name.inplace
```

Comment:
I'm not quite sure if I understand your question..
This function tells if `function_schema` is a native, non-foreach function which `f`, a foreach function in question, can reference to generate derivatives. with this in mind, it would be reasonable to filter out in-place native functions.
I have no idea how I should explicitly cope with inplace foreach functions

Reply:
Oh sorry I wasn't clear. Thanks for the clarification. I don't think we need to support inplace foreach with this PR, that is much trickier!